pyo3-macros-backend/src/utils.rs

```diff
@@ -137,17 +137,6 @@ impl quote::ToTokens for PythonDoc {
     }
 }
 
-pub fn ensure_not_async_fn(sig: &syn::Signature) -> syn::Result<()> {
```

Comment:
What is the error message if the optional dependency is disabled and an `async` function or method is used?

Reply:
You're right, I have not think about this use case. I will add a `compile_error`.

Reply:
Actually, I can't add a `compile_error!` simply because it's not possible to know at macro expansion if the feature is enabled or not.
So the error message will be that `_pyo3::impl_::coroutine` doesn't exists. I know that this error message is not very explicit, but because the feature activated by default, I don't think it really matter.

A solution could be to export a global declarative macro like this
```rust
#[cfg(feature = "async")]
macro_rules! check_async {
    ($($tokens:tt)*) => { 
        $($tokens)*
    };
}

#[cfg(not(feature = "async"))]
macro_rules! check_async {
    ($($tokens:tt)*) => { 
        compile_error!("async fn support requires pyo3 \"async\" feature")
    };
}
```
but do you think it's worth adding a global macro to check for a default feature?

Reply:
I think one way to solve this is add an `async` feature to `pyo3-macros-backend`. Can then add a friendly error message instructing to enable the feature if it's not enabled.

Reply:
This solution would bring the strange edge case of having it enabled on macro side but not on pyo3 side.

pyo3 has only 2 default features: `async` and `macros`. What could be the reason to disable default features, and only use `macros`? (I can see the interest of disabling macros only, but not the opposite) IMO this edge case doesn't really matter.

Otherwise, I still prefer a `#[doc(hidden)]` exported declarative macro.

Reply:
> pyo3 has only 2 default features: async and macros. What could be the reason to disable default features, and only use macros?

One thing that comes to mind is avoid the related optional dependencies. If we say that "futures-task just is not big enough to want to avoid it", then why have a feature to disable it in the first place. Otherwise, it should be possible to build PyO3 with the smallest possible dependency closure. (Especially for libraries which extend PyO3 itself and should not pollute the dependency management of their users.)

Reply:
My preference would be to just not have this async feature. I don't think futures is big enough to justify the added complexity.

Reply:
The feature is dropped